torch/_inductor/memory.py

```diff
@@ -107,63 +125,127 @@ def compute_size_for_scheduler_buffer(name_to_buf: Dict[str, SchedulerBuffer]) -
         buf0: at creation, 30 bytes allocated, when deleted, 0 bytes freed
         buf1: at creation, 0 bytes allocated, when deleted, 10 bytes freed
         buf2: at creation, 0 bytes allocated, when deleted, 20 bytes freed
+
+    Returns:
+        A dictionary mapping a scheduler buffer to a tuple of (size_alloc, size_free).
     """
-    from .scheduler import BaseSchedulerNode, OutputNode
+    from .ir import MultiOutput
+    from .scheduler import OutputNode
 
-    # compute the size of SchedulerBuffer without MultiOutputLayout layout
-    for sched_buf in name_to_buf.values():
-        if not isinstance(sched_buf.node.layout, MultiOutputLayout):
-            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
-            sched_buf.mpi_buffer.size_alloc = V.graph.sizevars.size_hint(
-                sched_buf.node.get_numel(), fallback=0
-            ) * get_dtype_size(sched_buf.node.get_dtype())
-            sched_buf.mpi_buffer.size_free = sched_buf.mpi_buffer.size_alloc
+    sched_buf_to_size: Dict[str, Tuple[int, int]] = dict()
 
-    # compute the size of SchedulerBuffer with MultiOutputLayout layout
-    for sched_buf in name_to_buf.values():
+    def _get_buf_size(
+        sched_buf: SchedulerBuffer, user_of_MultiOutputLayout: bool = False
+    ) -> int:
         if isinstance(sched_buf.node.layout, MultiOutputLayout):
-            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
+            size_alloc = 0
             for user in sched_buf.users:
                 if isinstance(user.node, OutputNode):
                     continue
-                assert isinstance(user.node, BaseSchedulerNode)
                 for buf in user.node.get_outputs():
-                    sched_buf.mpi_buffer.size_alloc += buf.mpi_buffer.size_alloc
-                    buf.mpi_buffer.size_alloc = 0
+                    if isinstance(buf.node, MultiOutput):
+                        size_alloc += _get_buf_size(buf, True)
+            sched_buf_to_size[sched_buf.get_name()] = (
+                0 if user_of_MultiOutputLayout else size_alloc,
+                0,
+            )
+            return size_alloc
+        else:
+            buf_size = V.graph.sizevars.size_hint(
+                sched_buf.node.get_numel(), fallback=0
+            ) * get_dtype_size(sched_buf.node.get_dtype())
+            sched_buf_to_size[sched_buf.get_name()] = (
+                0 if user_of_MultiOutputLayout else buf_size,
+                buf_size,
+            )
+            return buf_size
 
+    for sched_buf in name_to_buf.values():
+        # skip if sched_buf is already processed as an user of another SchedulerBuffer
+        # whose layout is of the type MultiOutputLayout
+        if sched_buf.get_name() not in sched_buf_to_size:
+            _ = _get_buf_size(sched_buf)
```

Comment:
Nit: maybe name it more transparently, `compute_and_update_buf_size` or something, up to you. and as far as i'm aware convention is just to call `_get_buf_size(sched_buf)` without assigning output to `_`. Also, fun fact - `_` is actually a live variable lol. so if you assign it something you want to have deallocated it will have been kept alive haha. Not that the latter part is particularly relevant here

Reply:
> _ is actually a live variable lol.

TIL, feels like I have been lied to by Python.